### ISSUE TITLE
Correct way to access want_capture_mouse and friends

### ISSUE THREAD

**Issue by robo-corg** [Gekkio/imgui-rs#128]
I really like this library and imgui. Thanks for writing this wrapper!

I read the FAQ in dear imgui and it mentions using want_capture_mouse etc.. for determining if input should be sent to imgui or the game, however those are only members of Ui which only has an immutable reference to ImGui.

What this means (I think) is you can't send input to ImGui then take that same event and use want_capture_mouse to try to decide if you are going to pass it on to the games code on the fly.

I can see a few workarounds here like buffering events or using the last frame's values for want_* but I am curious if there is a "right" way to do it.

Maybe we could update the examples to have this in them?

**Comment by thomcc** [comment]
Agreed, it's a big pain to access these where they are now, they should be on `ImGui`.

**Comment by robo-hamburger** [comment]
Yeah I added those in a similar way and it seems to be working great, as I feed ImGui glutin events I can filter those events out from ever getting to my game code. Maybe one of us should submit an MR.

**Comment by Gekkio** [comment]
The upstream API is a bit confusing here and there's a reason why those were put in Ui:

```
    //------------------------------------------------------------------
    // Output - Retrieve after calling NewFrame()
    //------------------------------------------------------------------

    bool        WantCaptureMouse;           // When io.WantCaptureMouse is true, do not dispatch mouse input data to your main application. This is set by ImGui when it wants to use your mouse (e.g. unclicked mouse is hovering a window, or a widget is active). 
    bool        WantCaptureKeyboard;        // When io.WantCaptureKeyboard is true, do not dispatch keyboard input data to your main application. This is set by ImGui when it wants to use your keyboard inputs.
    bool        WantTextInput;              // Mobile/console: when io.WantTextInput is true, you may display an on-screen keyboard. This is set by ImGui when it wants textual keyboard input to happen (e.g. when a InputText widget is active).
    bool        WantMoveMouse;              // [BETA-NAV] MousePos has been altered, back-end should reposition mouse on next frame. Set only when 'NavMovesMouse=true'.
    float       Framerate;                  // Application framerate estimation, in frame per second. Solely for convenience. Rolling average estimation based on IO.DeltaTime over 120 frames
    int         MetricsAllocs;              // Number of active memory allocations
    int         MetricsRenderVertices;      // Vertices output during last call to Render()
    int         MetricsRenderIndices;       // Indices output during last call to Render() = number of triangles * 3
    int         MetricsActiveWindows;       // Number of visible root windows (exclude child windows)
    ImVec2      MouseDelta;                 // Mouse delta. Note that this is zero if either current or previous position are invalid (-FLT_MAX,-FLT_MAX), so a disappearing/reappearing mouse won't have a huge delta.
```

> Retrieve after calling NewFrame()

It means that those variables are updated **after a frame has started** (in imgui-rs, this means after `frame` has been called and `Ui` is available). Dear ImGui also wants all information about input before starting the frame, so to me it sounds like this is what the main loop should look like:

1. dispatch input events to ImGui
2. begin ImGui frame
3. check want_capture_mouse and friends -> dispatch input events to your own app depending on the state
4. ...
5. finish ImGui frame and render it
6. goto 1

It's a bit confusing but that's what the upstream API suggests.

Now, we *could* move those functions to `ImGui`, but that means *you'll always read last frame's values* and all behaviour depending on those values will be off by one frame. For example, let's think about the upstream API `CaptureMouseFromApp` function: it's supposed to manually override the WantCaptureMouse for the next frame, but in this case it would take effect one frame too late.

Note that this also suggests that some existing functions I've added in the main `ImGui` struct are *wrong* (e.g. `frame_rate`) and read last frame's values...

**Comment by thomcc** [comment]
Ah, I see. This indicates to me that the lifetime bound on `Ui` is a problem, and it's causing a mismatch between how the library was intended to be used and how it realistically can be used.

As it is it heavily encourages you to limit the size of the scope where the `Ui` object is around, since having that borrow out is quite inconvenient, since it means you can't call any mutable methods on the whatever parent object holds onto it while the frame is going on.

Specifically, I suspect many uses of ImGui hold onto the ImGui object in a field in some struct in their applicatio state. E.g.

```rust
struct App {
    // stuff ...
    // Note: this might be nested or something, it doesn't matter.
    imgui: ImGui,
    // more stuff ...
}
```

The problem here is that if you do this, once you begin the ImGui frame, you can no longer update the state of `App`, since you have a borrow out on one of it's fields. So the loop you suggest is impossible for a design like this, which I suspect is probably common.

That is to say, I don't think the loop you suggest is possible unless you either

- Keep the ImGui object on the stack.
- Do something gross like

```rust
struct App {
    imgui: ImGui,
    rest_of_app_state: RefCell<RestOfAppState>,
}
```

Which is pretty awful to ask of users, especially when the lifetime bound is mostly artificial -- the underlying C++ API is just a bunch of free functions that you can call whenever (which, if I'm being honest, is probably what you want for a library intended for debug UI...).

**Comment by Gekkio** [comment]
Why does `ImGui` have to be part of App? What you're describing sounds like a very normal design challenge with Rust, which is trivial to fix with some restructuring, composition, and parameter passing. You can always go "up one layer" and use disjoint borrows:

```rust
struct CompositionIsYourFriend {
    imgui: ImGui,
    app: App,
}

struct App {
    very_important_state: u32,
}

impl App {
    fn run_frame(&mut self) {}
}

fn run_main_loop(mut c: CompositionIsYourFriend) {
    loop {
        let ui = c.imgui.frame((0, 0), (0, 0), 0.0);
        c.app.run_frame();
        draw_ui(&ui, &e.app);
        ui.render(|_, draw_list| -> Result<(), ()> {
            // use renderer
            Ok(())
        }).expect("Rendering failed");
    }
}
```

And if `App` needs to use some imgui-rs functions, pass Ui by reference to `run_frame`.

I used to hit this design challenge quite often in my own Rust code, until I realized the real problem was my way of designing things. With my heavy Java background I had a lot of design habits that are bad with Rust.

**Comment by thomcc** [comment]
Sure, but then you still have the restriction that you can't have any `mut` methods on `c` during the frame. This makes it hard for something to transparently use ImGui, and it also might require an annoying restructure to integrate into your code, which seems bad for a library that was originally designed for visualizations, prototypes, and debugging UIs.

It still seems needless to have these restrictions when they don't exist in the wrapped API (all the functions taking a `_: Ui<'a>` or methods on `Ui<'a>` that don't use self in their body...). It makes it much harder to use for debug UI and prototypes when compared to the C++ library (this is getting off topic for this issue though).

**Comment by Gekkio** [comment]
> It still seems needless to have these restrictions when they don't exist in the wrapped API

The goal of this library is to provide a safe wrapper for ImGui. If you don't want safety, use `imgui-sys`. If you have an alternative design for the entire library that still provides the same safety guarantees, I'm interested and you should open a new issue that describes it. I'm still a Rust beginner, so I know my design is not perfect. But I don't want to spend time debugging the debug UI of my app and don't want random process abortions, so I'll continue trying to maintain the safety guarantees of the main crate.

Now, about the original issue: having these functions on the `Ui` struct means that you can only use them correctly. The only thing adding them to `ImGui` accomplishes is incorrect usage patterns. Is this really what we want? If the upstream library updated these flags in Render/EndFrame, it would be ok. But as it stands, the correct values are updated after `Ui` is available during a frame. Remember, if you want to do the wrong (but "convenient") thing, you can always use `imgui-sys` and bypass all safety.

**Comment by robo-hamburger** [comment]
I see now the examples even call it out:   https://github.com/ocornut/imgui/blob/948009a8b2e98ef35fb8ddfe19299535a16c0834/examples/sdl_opengl3_example/imgui_impl_sdl_gl3.cpp#L495

I was trying to avoid having to buffer my events but it looks like I don't have any choice if I want to do it the right way.

The api you have with Ui seems reasonable enough. I think better docs & examples either upstream or in this library would have helped. Also me being less lazy 😄 

Maybe the solution for "I just want get ImGui running" up and running would be for a crate with a higher level api. I have effectively written that for my own game in a module called imgui_support.rs that was inspired by the examples in this repo.

**Comment by Gekkio** [comment]
I've changed how the upstream ImGuiIO struct is handled in the upcoming 0.1 release, and the rules have been relaxed a bit. Basically, instead of individual setter/getter functions, the IO struct is accessible via immutable reference directly from both `Context` (previously known as `ImGui`) and `Ui`. This means that you are basically free to read `want_capture_mouse` at any point you like. The upstream docs and examples are still confusing, but after looking at the C++ implementation my recommendation still stands: access those fields *only after* `frame()`. But now it's just a recommendation :wink: 

No ETA for the 0.1 release yet though, so I'll keep this issue open for now.

**Comment by tangmi** [comment]
This comment is just for consideration.

My setup is similar to below:

```rust
// imgui initialized up here

loop {
    events_loop.poll_events(|event| {
        imgui_winit_support::handle_event(
            &mut imgui,
            &event,
            window_hidpi_factor,
            window_hidpi_factor,
        );

        if let winit::Event::WindowEvent {
            event, ..
        } = event
        {
            let io = unsafe { &*imgui_sys::igGetIO() };

            match event
            {
                winit::WindowEvent::KeyboardInput {
                    ..
                } if io.want_capture_keyboard =>
                {
                    // ImGUI handled input
                }

                winit::WindowEvent::MouseInput {
                    ..
                }
                | winit::WindowEvent::MouseWheel {
                    ..
                } if io.want_capture_mouse =>
                {
                    // ImGUI handled input
                }

                _ => app.on_event(event),
            }
        }
    });

    // ui created down here, `app` updated and rendered
}
```

I'm unsafely getting the `IO` to see which events to ignore. Currently `imgui-rs` exposes this through `Ui`. If I create the `Ui` for the frame above the `poll_events` call, then I end up with a double borrow on `imgui` through `imgui.frame` and `imgui_winit_support::handle_event`. If the `Imgui` type exposed `want_capture_*`, I believe I would not need the unsafe access to `imgui-sys`.

Please let me know if I'm using imgui completely wrong and there's a way better way to do this!

**Comment by Gekkio** [comment]
Fixed by `0.1.0-rc.1`
